oauthenticator/google.py

```diff
@@ -389,20 +370,22 @@ def _fetch_member_groups(
         """
         Return a set with the google groups a given user/group is a member of, including nested groups if allowed.
         """
-        # FIXME: When this function is used and waiting for web request
-        #        responses, JupyterHub gets blocked from doing other things.
-        #        Ideally the web requests should be made using an async client
-        #        that can be awaited while JupyterHub handles other things.
-        #
-        if not hasattr(self, 'service'):
-            self.service = self._setup_service(user_email_domain, http)
+        if not hasattr(self, 'credentials'):
+            self.credentials = await self._setup_credentials(user_email_domain)
```

Comment:
I think this could be trouble, but I'm not sure.

`self` is an authenticator object, I'm not sure how long lived they are and if they are tied to a specific user. I think they are tied to a specific user, but I think they may be somewhat long lived, due to that, it could be that credentials is requested and saved but that they expire after a while, and then things go bad thereafter.

@minrk could you help us here, whats the lifecycle of an authenticator instance - and are they instanciated per user?

Reply:
Sounds like this alleviates @consideRatio's worry, but seems like I may need to add some logic to refresh the oauth token depending on how long-lived this object is?

Reply:
Ah no if the object is long lived, as demonstrated by @manics with count being incremented over time, it means that setting credentials will be done once and between users, and the'll expire etc.

I think the thing to do is to just ensure that you get credentials once per recursive-call-chain to the fetch member groups function, like is done for checked groups parameter - its initialized once, but wont be reused between users etc

Reply:
If the credentials used to fetch groups has been tied to a user, then we absolutely must re-acquire credentials and not re-use those.

Reply:
I believe my latest commit [7884369](https://github.com/jupyterhub/oauthenticator/pull/764/commits/7884369e7f16330ca1492d725c634f6d8b7b6116) adds what you're getting at here, @consideRatio. No longer an attrib of the object, it is refreshed for each login. 

According to [Google's oauth docs](https://developers.google.com/identity/protocols/oauth2#expiration) there is a limit to the number of refresh tokens before the oldest ones begin to be invalidated, but this limit does **not** apply to service accounts so I believe we've got everything resolved here.

I will test this with a group of users today.